modcc/printer/infoprinter.cpp

```diff
@@ -12,134 +18,147 @@
-        << name << "();\n"
-                   "\n"
-                   "inline const ::arb::mechanism_info& mechanism_"
-        << name << "_info() {\n"
-        << indent;
-
-    any_fields&& out << "using spec = ::arb::mechanism_field_spec;\n";
-
-    out << "static ::arb::mechanism_info info = {\n"
-        << indent << "// globals\n"
-                     "{\n"
-        << indent;
-
-    io::separator sep(",\n");
-    for (const auto& id: ids.global_parameter_ids) {
-        out << sep << id_field_info(id, "global");
-    }
-
-    out << popindent << "\n},\n// parameters\n{\n"
-        << indent;
+    std::string fingerprint = "<placeholder>";
 
-    sep.reset();
-    for (const auto& id: ids.range_parameter_ids) {
-        out << sep << id_field_info(id, "parameter");
+    out << fmt::format("#pragma once\n\n"
+                       "#include <cmath>\n"
+                       "#include <{}mechanism_abi.h>\n\n",
+                       arb_header_prefix());
+
+    auto vars = local_module_variables(m);
+    auto ion_deps = m.ion_deps();
+
+
+    std::unordered_map<std::string, Id> name2id;
+    for (const auto& id: m.parameter_block().parameters) name2id[id.name()] = id;
+    for (const auto& id: m.state_block().state_variables) name2id[id.name()] = id;
+
+    auto fmt_var = [&](const auto& v) {
+        auto kv = name2id.find(v->name());
+        auto lo = std::numeric_limits<double>::lowest();
+        auto hi = std::numeric_limits<double>::max();
+        std::string unit = "";
+        if (kv != name2id.end()) {
+            auto id = kv->second;
+            unit = id.unit_string();
+            if (id.has_range()) {
+                auto lo = id.range.first;
+                auto hi = id.range.second;
+            }
+        }
+        return fmt::format("{{ \"{}\", \"{}\", {}, {}, {} }}",
+                           v->name(),
+                           unit,
+                           std::isnan(v->value()) ? "NAN" : std::to_string(v->value()),
+                           lo, hi);
+    };
+
+    auto fmt_ion = [](const auto& i) { return fmt::format(FMT_COMPILE("{{ \"{}\", {}, {}, {}, {}, {}, {}, {}}}"),
+                                                          i.name,
+                                                          i.writes_concentration_int() ? "true" : "false",
+                                                          i.writes_concentration_ext() ? "true" : "false",
+                                                          i.writes_rev_potential() ? "true" : "false",
+                                                          i.uses_rev_potential() ? "true" : "false",
+                                                          i.uses_valence() ? "true" : "false",
+                                                          i.verifies_valence() ? "true" : "false",
+                                                          i.expected_valence); };
```

Comment:
Can use boolean values directly with `fmt`; no need for the `?:`s here. Also, a line break before `return` above would allow things to be less indented here, which would help keep things neater.

Reply:
👍